GiniVision/Classes/Core/Gallery/GalleryCoordinator.swift

```diff
@@ -206,13 +226,38 @@ extension GalleryCoordinator: ImagePickerViewControllerDelegate {
     }
     
     func imagePicker(_ viewController: ImagePickerViewController,
-                     didDeselectAsset asset: Asset) {
-        if let index = selectedImageDocuments.index(forKey: asset.identifier) {
-            selectedImageDocuments.remove(at: index)
+                     didDeselectAsset asset: Asset,
+                     at index: IndexPath) {
+        if let documentIndex = selectedImageDocuments.index(forKey: asset.identifier) {
+            viewController.deselectCell(at: index)
+            selectedImageDocuments.remove(at: documentIndex)
         }
         
-        if let selectedItems = viewController.collectionView.indexPathsForSelectedItems, selectedItems.isEmpty {
+        if selectedImageDocuments.isEmpty {
             viewController.navigationItem.setRightBarButton(cancelButton, animated: true)
         }
     }
+    
+    private func addSelected(_ asset: Asset, withData data: Data) {
+        var data = data
+        
+        // Some pictures have a wrong bytes structure and are not processed as images.
+        if !data.isImage {
+            if let image = UIImage(data: data),
+                let imageData = UIImageJPEGRepresentation(image, 1.0) {
+                data = imageData
+            }
+        }
+        let imageDocument = GiniImageDocument(data: data,
+                                              imageSource: .external,
+                                              imageImportMethod: .picker,
+                                              deviceOrientation: nil)
+        self.selectedImageDocuments[asset.identifier] = imageDocument
+        
+        if !self.giniConfiguration.multipageEnabled {
+            DispatchQueue.main.async {
```

Comment:
It is something that really displeases me, and I think the best solution as you suggest is to split into different methods.

Reply:
This is something that really displeases me, I like your proposal of splitting them into methods and I think that it will fit perfectly with the likely new approach for concurrency that will come in Swift 5 ([This](https://gist.github.com/lattner/31ed37682ef1576b16bca1432ea9f782))